ostd/src/mm/kspace/kva.rs

```diff
@@ -0,0 +1,362 @@
+        });
+        // 1. get the previous free block, check if we can merge this block with the free one
+        //     - if contiguous, merge this area with the free block.
+        //     - if not contiguous, create a new free block, insert it into the list.
+        let mut free_range = range.clone();
+
+        if let Some((prev_va, prev_node)) = freelist
+            .upper_bound_mut(core::ops::Bound::Excluded(&free_range.start))
+            .peek_prev()
+        {
+            if prev_node.block.end == free_range.start {
+                let prev_va = *prev_va;
+                free_range.start = prev_node.block.start;
+                freelist.remove(&prev_va);
+            }
+        }
+        freelist.insert(free_range.start, KvaFreeNode::new(free_range.clone()));
+
+        // 2. check if we can merge the current block with the next block, if we can, do so.
+        if let Some((next_va, next_node)) = freelist
+            .lower_bound_mut(core::ops::Bound::Excluded(&free_range.start))
+            .peek_next()
+        {
+            if free_range.end == next_node.block.start {
+                let next_va = *next_va;
+                free_range.end = next_node.block.end;
+                freelist.remove(&next_va);
+                freelist.get_mut(&free_range.start).unwrap().block.end = free_range.end;
+            }
+        }
+    }
+}
+
+pub static KVA_TRACKED_ALLOCATOR: VirtAddrAllocator =
+    VirtAddrAllocator::new(TRACKED_MAPPED_PAGES_RANGE);
+pub static KVA_UNTRACKED_ALLOCATOR: VirtAddrAllocator = VirtAddrAllocator::new(VMALLOC_VADDR_RANGE);
+
+pub struct Tracked;
+pub struct UnTracked;
+
+pub trait AllocatorSelector {
+    fn select_allocator() -> &'static VirtAddrAllocator;
+}
+
+impl AllocatorSelector for Tracked {
+    fn select_allocator() -> &'static VirtAddrAllocator {
+        &KVA_TRACKED_ALLOCATOR
+    }
+}
+
+impl AllocatorSelector for UnTracked {
+    fn select_allocator() -> &'static VirtAddrAllocator {
+        &KVA_UNTRACKED_ALLOCATOR
+    }
+}
+
+pub struct Kva<M: AllocatorSelector + 'static> {
+    range: Range<Vaddr>,
+    num_cpus_: u32,
+    phantom: PhantomData<M>,
+}
+
+impl<M: AllocatorSelector + 'static> Kva<M> {
+    pub fn new(size: usize) -> Self {
+        let allocator = M::select_allocator();
+        let range = allocator.alloc(size).unwrap();
+        let num_cpus_ = num_cpus();
+        Self {
+            range,
+            num_cpus_,
+            phantom: PhantomData,
+        }
+    }
+
+    pub fn start(&self) -> Vaddr {
+        self.range.start
+    }
+
+    pub fn end(&self) -> Vaddr {
+        self.range.end
+    }
+
+    pub fn range(&self) -> Range<Vaddr> {
+        self.range.start..self.range.end
+    }
+
+    /// Map pages into the kernel virtual area.
+    /// # Safety
+    /// The caller should ensure either the mapped pages or the range to be used doesn't
+    /// violate the memory safety of kernel objects.
+    pub unsafe fn map_pages<T: PageMeta>(
```

Comment:
Since the range is allocated, we can now ensure that the mapped range is valid. Since the typed KVA does not provide accessing methods for the data. You can remove the `unsafe` mark and the requirement of this function.